### PR TITLE
Add support for cracking IBM 4690 ADXCRYPT hashes

### PR DESCRIPTION
**Pull request by kholia**
Note: I will be tuning the format properly soon.

Update: Done now.

### CODE REVIEW

**Comment by kholia** [comment]
https://github.com/kholia/4690-OS has some more information on this topic.

File: src/adxcrypt_fmt_plug.c [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+ * Known plain text = 30
+ * Cipher text = 43891846
+ *
+ * Known plain text = 31
+ * Cipher text = 43691826
+ *
+ * Known plain text = 32
+ * Cipher text = 43391806
+ */
+static int adxcrypt(char *input, unsigned char *output, int16_t length)
```
[the remark this answers is not in the document]
Comment (kholia) [reply src/adxcrypt_fmt_plug.c]: Ack. Will fix this. Thanks!

**Comment by kholia** [comment]
An OpenCL format would be an overkill :)

The CPU format finds usable collisions pretty quickly (within minutes).

**Comment by magnumripper** [comment]
In IBM we trust.

**Comment by magnumripper** [comment]
What the heck is this anyway, some 1980's format? Or is it actually in use at my bank right now?!?

**Comment by kholia** [comment]
IBM/Toshiba 4690 is an OS for POS (point of sale) machines. It is from the 70's but is still being developed and used today.

I know very little about it.

**Comment by magnumripper** [comment]
Fantastic. Management by not having any clue whatsoever.

**Comment by kholia** [comment]
https://github.com/kholia/4690-OS has info on getting the 4690 OS running locally.

It combines DOS, Linux, and IBM's Mainframe styles screens :)

**Comment by claudioandre-br** [comment]
> Fantastic. Management by not having any clue whatsoever.

Indeed. And the companies/teams that use this (kind of software) generally love to talk about synergy, management, blablabla, but do not invest in training, research (they do not even try to get clues in JtR, Hashcat, ...).

**Comment by magnumripper** [comment]
Just hoping I'm not ending up collateral damage after approving this PR 😆 🤣

**Comment by magnumripper** [comment]
Apparently Toshiba spent some 40 man-hours of development budget in finding a nice background screen. They could've have spent it in a security audit, but hey that's very boring.

**Comment by magnumripper** [comment]
> "Thanks to Brian Kamusinga and David Odell for reporting this vulnerability."

Wow, they really didn't have a clue until someone told them. Amazing!

**Comment by kholia** [comment]
On a 400$ Chromebook from 2014 with i3-4005U CPU.

```
$ ../run/john --test --format=adxcrypt 
Will run 4 OpenMP threads
Benchmarking: adxcrypt [IBM/Toshiba 4690 - ADXCRYPT 32/64]... (4xOMP) DONE
Many salts:	26473K c/s real, 9583K c/s virtual
Only one salt:	25690K c/s real, 9479K c/s virtual
```

**Comment by kholia** [comment]
https://github.com/kholia/4690-OS/tree/master/data has sample data -> ADXCSOUF.DAT and ADXEPW0F.DAT files.

**Comment by claudioandre-br** [comment]
> Apparently Toshiba spent some 40 man-hours of development budget in finding a nice background screen. They could've spent it in a security audit, but hey that's very boring.

It is very sad to see how "business" chooses its priorities.

**Comment by magnumripper** [comment]
We could write a SIMD-version of the adxcrypt algorithm, just for fun. Looks like it'd be trivial. But it's so fast anyway it's probably not really worth the effort.

**Comment by kholia** [comment]
Yeah, it might be an overkill :-)

**Pull request merged**